dayon: make deterministic and clean up

TomaSajt opened the pull request:
## Description of changes

Changes:
- use `finalAttrs` instead of `rec`
- use `canonicalize-jars-hook` (related issue: https://github.com/NixOS/nixpkgs/issues/278518)
- patch `.desktop` files to find icons
- add `meta.mainProgram`


<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

fgaz commented on pkgs/applications/networking/remote/dayon/default.nix:
```diff
@@ -32,15 +32,15 @@ stdenv.mkDerivation rec {
+      --replace "/usr/bin/dayon/dayon.png" "dayon"
+
     install -Dm644 build/dayon.jar $out/share/dayon/dayon.jar
-    mkdir -p $out/bin
```
TIL that makeWrapper creates parent directories

fgaz reviewed (approved):
Looks good, thank you

fgaz commented on pkgs/applications/networking/remote/dayon/default.nix:
```diff
@@ -32,15 +32,15 @@ stdenv.mkDerivation rec {
-    "resources/deb/dayon_assistant.desktop"
-  ];
-
   installPhase = ''
     runHook preInstall
+
+    install -Dm644 resources/dayon.png $out/share/icons/hicolor/128x128/apps/dayon.png
+    install -Dm644 resources/deb/*.desktop -t $out/share/applications
+    substituteInPlace $out/share/applications/*.desktop \
+      --replace "/usr/bin/dayon/dayon.png" "dayon"
+
```
On second thought, can we leave copyDesktopItems and patch the desktop files in pre/postFixup? You're copying them as is and modifying them later anyway

TomaSajt replied:
Okay, done

TomaSajt committed:
dayon: make deterministic and clean up

Changes:
- use `finalAttrs` instead of `rec`
- use `canonicalize-jars-hook` (related issue: https://github.com/NixOS/nixpkgs/issues/278518)
- patch `.desktop` files to find icons
- add `meta.mainProgram`

TomaSajt commented:
> > Changes:
> > 
> > * use finalAttrs instead of rec
> > * use canonicalize-jars-hook (related issue:  [Tracking issue: non-deterministic java apps #278518](https://github.com/NixOS/nixpkgs/issues/278518))
> > * manually install .desktop files, because the icon paths were broken
> > * add meta.mainProgram
> 
> IMHO, this should be in the commit message body.

Good idea. I never really used commit message bodies previously, maybe it's time to start
